R/defineFeasibleSet.R

```diff
@@ -1,20 +1,179 @@
+#'   model = mod,
+#'   frac.within = 0.9, within.rel = 0.15, within.abs = 1
+#' )
+#'
+#' ## How many of the 300 possible parameter sets were retained?
+#' nrow(coef(feas, feasible.set = TRUE))
+#'
+#' ## View ranges of parameters in feasible set
+#' feas
+#'
+#' ## Plot simulation bounds
+#' xyplot(feas, feasible.bounds = TRUE, cut = 3)
+#'
+#' ## Generate set of simulations with NSE > 0.5, for GLUE.
+#' ## First, need to calculate objective function values:
+#' fit <- fitBySampling(mod, samples = 300, objective = hmadstat("r.squared"))
+#' ## Calculate 5 percent and 95 percent GLUE quantiles (i.e. weighted).
+#' fitglu <- defineFeasibleSet(fit,
+#'   threshold = 0.5,
+#'   glue.quantiles = c(0.05, 0.95)
+#' )
+#' ## Coverage of the GLUE quantile simulations (within 0.1 mm/day)
+#' sim <- fitted(fitglu, feasible.bounds = TRUE)
+#' head(sim)
+#' mean((sim[, 1] < observed(fitglu) + 0.1) &
+#'   (sim[, 2] > observed(fitglu) - 0.1))
+#'
+#' ## Or - keep adding parameter sets until we reach a target coverage:
+#' ## Calculate 5 percent and 95 percent GLUE quantiles (i.e. weighted).
+#' fitglu <- defineFeasibleSet(fit,
+#'   target.coverage = 0.9,
+#'   glue.quantiles = c(0.05, 0.95)
+#' )
+#' ## Coverage of the GLUE quantile simulations (within 0.1 mm/day)
+#' ## (not to be confused with the target.coverage to define overall feasible set)
+#' sim <- fitted(fitglu, feasible.bounds = TRUE)
+#' mean((sim[, 1] < observed(fitglu) + 0.1) &
+#'   (sim[, 2] > observed(fitglu) - 0.1))
+#'
+#' ## Plot simulated GLUE quantiles
+#' xyplot(fitglu, feasible.bounds = TRUE, cut = 3)
+#' xyplot(fitglu,
+#'   feasible.bounds = TRUE, cut = 3,
+#'   scales = list(y = list(log = TRUE))
+#' )
+#' ## Summarise size of the simulation bounds: lower as fraction of upper
+#' summary(coredata(sim[, 1] / sim[, 2]))
+#'
+#' ## Simulate on a new data period
+#' newglu <- update(fitglu,
+#'   newdata = window(Queanbeyan,
+#'     start = "1980-01-01", end = "1982-01-01"
+#'   ),
+#'   glue.quantiles = c(0.05, 0.95)
+#' )
+#' ## The new period is very dry, all model simulations overestimate flow.
+#' xyplot(newglu, feasible.bounds = TRUE, cut = 3)
+#' ## Coverage of the GLUE quantile simulations (within 0.1 mm/day)
+#' sim <- fitted(newglu, feasible.bounds = TRUE)
+#' mean((sim[, 1] < observed(newglu) + 0.1) &
+#'   (sim[, 2] > observed(newglu) - 0.1))
+#' @export
 defineFeasibleSet <- function(x, ...) {
   UseMethod("defineFeasibleSet")
 }
 
+
+#' @rdname defineFeasibleSet
+#' @export
 defineFeasibleSet.hydromad <-
   function(x, ..., thin = NA) {
     x <- update(x, feasible.set = NULL)
     if (inherits(x$fit.result, "dream")) {
       result <- x$fit.result
       ## extract last half of Sequences (assumed to have converged)
       mcmc <- window(result)
-      logp <- as.mcmc(window(result$hist.logp, start = start(mcmc)))
+      logp <- coda::as.mcmc(window(result$hist.logp, start = start(mcmc)))
       ## thin the sequences to remove autocorrelation (more efficient samples)
```

Comment:
It's ok to make coda a suggests because this is not a core hydromad function - it relies on dream, which is suggested anyway.